### ISSUE TITLE
Go spike for the directory.

### ISSUE THREAD

**Issue by futurechimp** [nymtech/nym#2022]

**Comment by jstuczyn** [comment]
My goals for this investigation were to check whether the following are possible:
- [x] IMPLEMENTED: Ability to send a bond-mixnode tx that would save some basic information to the chain + ability to query for it
- [x] NOT IMPLEMENTED, BUT CONVINCED IT'S POSSIBLE: Extending the above with a way of sending tokens alongside it for the purposes of locking them
- [x] NOT IMPLEMENTED, BUT CONVINCED IT'S POSSIBLE: Getting back tokens with an unbond tx
- [x] IMPLEMENTED: Something (like this Osmosis' epoch, maybe we could import that?) that runs every X blocks or every X minutes that would increase mixnode's bond by X%; the unboding, etc should still work. 

in general, with the transitioning (or not) our mixnet functionalities from cosmwasm smart contract into a proper cosmos-sdk module, I can see the following pros and cons:

pros:
- removal of the external entity driving our epochs forward (i.e. validator API) and performing operations on its own internal timer, i.e. all the calls to "reconcile delegations", "reward mixnode", etc. -> as a result, removal of point of failure, more decentralisation + no need for paying fees for those transactions
- we could simplify our logic in multiple places as we could assume some things are actually atomic (like all of the operations happening every epoch). I'm not 100% sure, but we might also be able to get rid of all those "pending" events.
- no longer being limited by the block time when executing our code, i.e. currently we can't have "complex" transactions in our smart contract as it would fail to get executed within provided time (which is independent of set gas limit). For example even with infinite gas limits we were unable to eagerly reward all delegators. I'm not saying we want to transition to eager rewarding, I'm saying that we wouldn't be as limited. (I put a time.Sleep(15s) in cosmos-sdk during tx execution. The tx got accepted and the chain didn't halt)
- this is purely speculative, but I've got a feeling that distribution of generation of the active set or introduction of the mythical "random beacon" is going to be an order of magnitude easier (and less hacky) to implement inside a cosmos-sdk module as opposed to validator api + cosmwasm smart contract

cons:
- very time consuming to implement and migrate. there's a non-zero chance it's going to be actually impossible with the existince of our vesting contract and its interaction with the mixnet contract
- bugs in this code could now potentially become chain-halting

tl;dr: it's risky but it might clean up bunch of things and get rid of so many workarounds we have in our contract to deal with its limitations
